### ISSUE TITLE
Location drop-downs for multiple selections

### ISSUE THREAD

**Issue by javedyousafzai** [humanitarianresponse/site#96]
This is already available in the atrium case case which was originally asked by Christian Habib from Syria regional office. We had several suggestions to solve this but none has been implemented in HR.info.

Atrium case: https://atrium.humanitarianresponse.info/humanitarian_response/node/4324

**Comment by humphreymburu** [comment]
@javedyousafzai 
I Will work on this.

**Comment by pcambra** [comment]
What about something like this http://www.uxbooth.com/articles/designing-a-progressive-location-filter/

**Comment by javedyousafzai** [comment]
yes, @andrejverity and I were talking about something similar. I am going to draw a basic mockup and will ask your input on the 'performance' side of it.

**Comment by javedyousafzai** [comment]
See the screen below on how the admin levels are associated with each other;

![image](https://cloud.githubusercontent.com/assets/3966785/10139873/49dcc774-6605-11e5-9f29-aacf1d1bca70.png)

Example: Let’s say, an IMO wants to add an assessment which is happening in a district ‘Asadabad’. This district actually belongs to Province ‘Kunar’ which belongs to the region ‘Eastern’.

![image](https://cloud.githubusercontent.com/assets/3966785/10139902/6db6557a-6605-11e5-9770-01838e76cf9a.png)

The ‘multiple location’ widget should works in the following manner;
1. First time when the form is loaded, the use will select the country ‘Afghanistan’
2. Then there will be a searchable ‘text input parameter’
3. As soon as IMO start typing ‘asa’, all the district containing ‘Asa’ should appear with the region name and province name preceding and then the actual district name – see the screenshot below.

![image](https://cloud.githubusercontent.com/assets/3966785/10139936/8ec25228-6605-11e5-9d7f-900893ae3ce4.png)

4. The searchable ‘text input parameter’ should accept multiple values as below;
![image](https://cloud.githubusercontent.com/assets/3966785/10139957/a26ccbc8-6605-11e5-94be-ffff96d8f37a.png)

Using the above widget, the user will have to know which district, state or region they want to tag for the content, but we expect them to know as they are working in that geographic area/country.

@pcambra! Now the questions is related to performance in term of querying the database to search for associated admin levels. Do you think its going to be a heavy to go through the location taxonomy (database) while adding the 'exact' location and show to the user the association between those admin levels?

**Comment by pcambra** [comment]
Having into account that we have over different 55000 locations, such a widget would be probably very slow. Generating all the combinations of Region-admin1\Province-admin2\District-admin3 and search for them in autocomplete, could be quite exhaustive.

**Comment by andrejverity** [comment]
@pcambra that 55,000 is the total number of locations we have in the system - corrected?  Two follow-up questions:

1) If we force the user to pre-filter by country, would that then no reduce the number of values we  have to search through (in the auto-complete) significantly?  And, if we were inside an Operation, we could already set the selected Country for the user.

2) If we generated combinations of strings, it should not increase the number of strings we have to search through - correct?

I was just chatting with Javed about the problem we are trying to solve (i.e. allowing people the ability to quickly tag MULTIPLE locations) so we will have some more questions & ideas coming soon.

**Comment by javedyousafzai** [comment]
@pcambra see this [assessment] (https://www.humanitarianresponse.info/en/operations/nepal/assessment/initial-rapid-assessment-ira/edit) as an example foom Nepal to show how many location a user can tag and how challenging it gets.. 
![image](https://cloud.githubusercontent.com/assets/3966785/10157932/a115bc1a-668e-11e5-90e5-877de830fbb9.png)

Another idea we are thinking of is to show all the locations hierarchy drop-downs filled in based on the previous selection when the user hit 'Add another item. This is ideal solution if the user is selecting multiple lowest (lowest level) in the same province and will will reduce number of clicks. See the screenshot below highlighted in 'red'. 

For example, I am selecting five districts from 'Kabul' state and I selected the first one. To select other four districts, I expect the site to show me the locations prepopulated based on previous selection

![image](https://cloud.githubusercontent.com/assets/3966785/10158354/cd9b5bca-6691-11e5-89e1-180fb8a2d9bc.png)

**Comment by pcambra** [comment]
@javedyousafzai that was my exact proposal yesterday on the prepopulation in the call. The issue I see with that approach is that if I add another item and change my mind, I need to manually empty the selectors, not the best option.
Maybe we can have two buttons, one to add a new item and another labeled as "Copy previous location" will behave as described above.

**Comment by andrejverity** [comment]
Two quick questions:
1) If "add" pre-populates and I want to choose somewhere else, can I simply just choose "-None-" in the country list and have everything else then removed for that row?
2) Also, perhaps a 'Clear' or 'X' (remove) button could help.

My thoughts are that if we can do the "copy" idea, that would already be a big step in helping people who are having to add multiple locations. It would buy us time to think about and find an even better solution.

What do you think @javedyousafzai ?

**Comment by andrejverity** [comment]
Does not seem to be supported, but something like this might have been nice: https://www.drupal.org/project/taxonomy_tree_select

**Comment by pcambra** [comment]
1) Yes, but I don't think is an obvious option for the user.
2) Yes, but we end up with a really complicated custom widget.

This is the alternative for D7 of that module: https://www.drupal.org/project/issues/term_reference_tree but it loads the terms all at once so it has performance issues (as reported in the issue queue)

We're using https://www.drupal.org/project/shs but seems there's an additional module that helps with performance that we're not using: https://www.drupal.org/project/js

**Comment by humphreymburu** [comment]
I'm looking into this options https://www.drupal.org/project/autocomplete_deluxe, https://www.lullabot.com/articles/entityreference-multiselectors seems interesting to.

**Comment by javedyousafzai** [comment]
@pcambra if we can test some viable solution based on above ideas/suggestions in our test environment, we can then review it. thanks

**Comment by Yaelle01** [comment]
Hi @pcambra , I found three fields that could be changed to auto-complete fields. Please find the screenshot below. Also, I am wondering how the contact reference is treated in these instances (re HID). I've equally marked the referring field to it. 

![assessments_form](https://cloud.githubusercontent.com/assets/13348894/11337546/21f31f82-91ee-11e5-9428-9dabe750a123.png)

**Comment by pcambra** [comment]
Related issue regarding the contact field: #195

**Comment by pcambra** [comment]
@Yaelle01 Themes field is a select list also in Event, Document, Map/Infographic, Contact and Blog post, shall we change it to autocomplete there too?
Same goes for Emergencies (which is labeled as Disaster in some content types)
